docs/stake-optimizer.md

```diff
@@ -0,0 +1,260 @@
+
+```python
+from cosmpy.aerial.client.distribution import create_withdraw_delegator_reward
+from cosmpy.aerial.client.staking import create_delegate_msg
+from cosmpy.aerial.tx import SigningCfg
+from cosmpy.aerial.wallet import LocalWallet
+from cosmpy.crypto.keypairs import PrivateKey
+from cosmpy.crypto.address import Address
+from cosmpy.aerial.tx import Transaction
+
+# Use any address with tokens available
+key = PrivateKey("XZ5BZQcr+FNl2usnSIQYpXsGWvBxKLRDkieUNIvMOV7=")
+alice = LocalWallet(key)
+alice_address = Address(key)._display
+
+tx = Transaction()
+
+# Add delegate msg
+tx.add_message(create_delegate_msg(alice_address,validator.address,initial_stake,"atestfet"))
+
+# Add claim reward msg
+tx.add_message(create_withdraw_delegator_reward(alice_address, validator.address))
+
+account = ledger.query_account(alice.address())
+tx.seal(SigningCfg.direct(alice.public_key(), account.sequence),fee="",gas_limit=0)
+tx.sign(alice.signer(), ledger.network_config.chain_id, account.number)
+tx.complete()
+
+# simulate the fee for the transaction
+_, str_tx_fee = ledger.estimate_gas_and_fee_for_tx(tx)
+```
+Since the output of this function is a string, we will convert it to an int and round it up to get a more conservative estimate for the `fee`
+
+```python
+denom = "atestfet"
+tx_fee = str_tx_fee[:-len(denom)]
+
+# Round up
+fee = round(int(tx_fee), -len(tx_fee)+1)
+```
+### Query Network Variables
+There are three network variables that we need to query since they will contribute to the staking rewards calculation: `total supply`, `inflation` and `community tax`
+
+```python
+# Total Supply of tokens
+req = QueryTotalSupplyRequest()
+resp = ledger.bank.TotalSupply(req)
+totalSupply = float(json.loads(resp.supply[0].amount))
+
+# Inflation
+req = QueryParamsRequest(subspace="mint", key="InflationRate") 
+resp = ledger.params.Params(req)
+inflation = float(json.loads(resp.param.value))
+
+# Community Tax
+req = QueryParamsRequest(subspace="distribution", key="communitytax") 
+resp = ledger.params.Params(req)
+communityTax = float(json.loads(resp.param.value))
+```
+
+## Calculate Reward Rate
+
+We can now proceed to calculate a theoretical staking rewards rate using the variables gathered above. These are: `inflation`, `totalSupply`, `pctDelegatedOfTotalStake`, `communityTax` and `commission`
+
+```python
+
+# Calculate anual reward
+anualDelegatorReward = (inflation * totalSupply) *pctDelegatedOfTotalStake* (1- communityTax)*(1- commission)
+
+# Convert from anual reward to minute reward
+minuteReward = anualDelegatorReward/360/24/60
+
+# Set the rate
+rate = minuteReward/initial_stake
+```
+
+## Calculate Optimal Compounding Period
+
+We can calculate the optimal compounding period that maximizes our staking rewards analytically by using the following function.
```

Comment:
Is there a reference for this that we could cite here?

Reply:
Not really, I could probably show the steps to get to that function, but I didn't really used any resources